### PR TITLE
Add support for QNN stateful models

### PR DESCRIPTION
**Pull request by qti-ashimaj**
This PR enhances support for QNN stateful models that execute entirely on Qualcomm's Neural Processing Unit (NPU), including attention operations and internal KV cache management. This significant improvement eliminates previous limitations where certain operations would fall back to CPU execution, resulting in performance bottlenecks.

The implementation identifies QNN stateful models through the `genie_model=True` provider option. For these models:
- The ONNX Runtime GenAI framework delegates KV cache management entirely to the QNN backend
- The model architecture is optimized to return logits only for the last token in the sequence

### CODE REVIEW

File: src/models/logits.cpp [hunk 1]
```diff
@@ -17,8 +18,8 @@ Logits::Logits(State& state)
 
   input_sequence_lengths.resize(state_.params_->search.batch_size);
 
-  if (IsOpenVINOStatefulModel(state.model_) || state.model_.IsPruned()) {
-    // In the case of OpenVINO stateful models, or any model whose ONNX graph
+  if (IsOpenVINOStatefulModel(state.model_) || IsQNNStatefulModel(state.model_) || state.model_.IsPruned()) {
```
Comment (baijumeswani) [inline src/models/logits.cpp]: Is the check for `state.model_.IsPruned()` not enough?

File: src/qnn/interface.cpp [hunk 2]
```diff
@@ -78,4 +79,24 @@ DeviceInterface* GetQNNInterface() {
   return g_device.get();
 }
 
+bool IsQNNStatefulModel(const Model& model) {
+  if (model.p_device_->GetType() == DeviceType::QNN || model.p_device_->GetType() == DeviceType::CPU) {
```
Comment (baijumeswani) [inline src/qnn/interface.cpp]: Why would p_device be CPU for QNN models that are stateful?

File: src/qnn/interface.cpp [hunk 3]
```diff
@@ -78,4 +79,24 @@ DeviceInterface* GetQNNInterface() {
...
+    const auto& provider_options = model.config_->model.decoder.session_options.provider_options;
+    for (auto& po : provider_options) {
```
Comment (baijumeswani) [inline src/qnn/interface.cpp]:
```suggestion
    for (const auto& po : provider_options) {
```

File: src/qnn/interface.cpp [hunk 4]
```diff
@@ -78,4 +79,24 @@ DeviceInterface* GetQNNInterface() {
...
+      if (po.name == "QNN") {
```
Comment (baijumeswani) [inline src/qnn/interface.cpp]: Does this need to handle other names? Like QNNExecutionProvider or qnn? Is this the standard naming for the qnn provider?

File: src/qnn/interface.cpp [hunk 5]
```diff
@@ -78,4 +79,24 @@ DeviceInterface* GetQNNInterface() {
...
+        const auto& qnn_options = po.options;
+        for (auto& option : qnn_options) {
```
Comment (baijumeswani) [inline src/qnn/interface.cpp]:
```suggestion
        for (const auto& option : po.options) {
```

File: src/qnn/interface.cpp (hunk 2 above)
Comment (baijumeswani) [reply src/qnn/interface.cpp]: Is it because npu memory is host accessible and that the memory allocations still happen on cpu?

Comment (baijumeswani) [reply src/qnn/interface.cpp]: If so, please add a comment.

File: src/models/logits.cpp (hunk 1 above)
Comment (qti-ashimaj) [reply src/models/logits.cpp]: Yes, we can use that API, removed the extra check

File: src/qnn/interface.cpp (hunk 2 above)
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: Device type is [QNN ](https://github.com/microsoft/onnxruntime-genai/blob/main/src/models/model.cpp#L751)only when we use the provider option `enable_htp_shared_memory_allocator=1`, Since we can run the stateful model regardless of the shared memory allocator, hence the checks for both device types.

File: src/qnn/interface.cpp (hunk 5 above)
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: Done

File: src/qnn/interface.cpp (hunk 4 above)
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: The normalized provider name is [QNN](https://github.com/microsoft/onnxruntime-genai/blob/main/src/config.cpp#L22) so we do not need to handle other strings.

File: src/qnn/interface.cpp [hunk 6]
```diff
@@ -78,4 +79,28 @@ DeviceInterface* GetQNNInterface() {
...
+        for (const auto& option : po.options) {
+          // For QNN, if session option 'genai_model' is set, the session will encapsulate
+          // a stateful model, so KVCache will be managed internally.
+          if (option.first == "genai_model" && option.second == "True") {
+            return true;
+          }
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: This check is only relevant for QNN models, hence it should rightly return false for non-CPU/non-QNN device type. The check is valid.

File: src/qnn/interface.cpp [hunk 7]
```diff
@@ -78,4 +79,28 @@ DeviceInterface* GetQNNInterface() {
...
+  // For QNN models without shared memory allocator, p_device_ will be CPU type
+  // Both cases need to be handled the same way for stateful models where KV cache is managed internally
+  if (model.p_device_->GetType() == DeviceType::QNN || model.p_device_->GetType() == DeviceType::CPU) {
+    const auto& provider_options = model.config_->model.decoder.session_options.provider_options;
+    for (const auto& po : provider_options) {
+      if (po.name == "QNN") {
+        for (const auto& option : po.options) {
+          // For QNN, if session option 'genai_model' is set, the session will encapsulate
+          // a stateful model, so KVCache will be managed internally.
+          if (option.first == "genai_model" && option.second == "True") {
```
[the remark this answers is not in the document]
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: Changed it to first convert the value to lower case and then compare with `\"true\"`

File: src/qnn/interface.h [hunk 8]
```diff
@@ -5,4 +5,7 @@ namespace Generators {
 
 DeviceInterface* GetQNNInterface();
 
+struct Model;
```
[the remark this answers is not in the document]
Comment (qti-ashimaj) [reply src/qnn/interface.h:8]: It is declared as [struct Model](https://github.com/microsoft/onnxruntime-genai/blob/main/src/models/model.h#L145) in `model.h`

File: src/qnn/interface.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/qnn/interface.cpp]:
```cpp
for (const auto& po : provider_options) {
      if (po.name == "QNN") {
        for (const auto& option : po.options) {
          // For QNN, if session option 'genai_model' is set, the session will encapsulate
          // a stateful model, so KVCache will be managed internally.
          if (option.first == "genai_model" && option.second == "True") {
            return true;
          }
        }
      }
    }
```

Already checks for the provider option. Do we need a check at the top as well? Because that check will still fail if the p_device = CPU (which is the case for multiple providers (such as QNN, VitisAI...).

[the remark this answers is not in the document]
Comment (qti-ashimaj) [reply src/qnn/interface.cpp]: Okay, I am removing the checks for device type. It will check for only QNN as provider followed by "genie_model"="true" as provider option. This will run regardless of device types.

**Comment by qti-ashimaj** [comment]
Hi @baijumeswani, could you please help in reviewing and getting the PR merged.
Thanks!
